astropy/stats/lombscargle/implementations/main.py

```diff
@@ -0,0 +1,227 @@
+"""
+Main Lomb-Scargle Implementation
+
+The ``lombscargle`` function here is essentially a sophisticated switch
+statement for the various implementations available in this submodule
+"""
+from __future__ import print_function, division
+
+__all__ = ['lombscargle', 'available_methods']
+
+import warnings
+
+import numpy as np
+
+from .slow_impl import lombscargle_slow
+from .fast_impl import lombscargle_fast
+from .scipy_impl import lombscargle_scipy
+from .chi2_impl import lombscargle_chi2
+from .fastchi2_impl import lombscargle_fastchi2
+from .cython_impl import lombscargle_cython
+
+
+METHODS = {'slow': lombscargle_slow,
+           'fast': lombscargle_fast,
+           'chi2': lombscargle_chi2,
+           'scipy': lombscargle_scipy,
+           'fastchi2': lombscargle_fastchi2,
+           'cython': lombscargle_cython}
+
+
+def available_methods():
+    methods = ['auto', 'slow', 'chi2', 'cython']
+
+    # Numpy 1.8 or newer required for fast algorithms
+    if hasattr(np.ufunc, 'at'):
+        methods.extend(['fast', 'fastchi2'])
+
+    try:
+        import scipy
+        methods.append('scipy')
+    except ImportError:
+        pass
+    return methods
+
+
+def _get_frequency_grid(frequency, assume_regular_frequency=False):
+    """Utility to get grid parameters from a frequency array
+
+    Parameters
+    ----------
+    frequency : array_like or Quantity
+        input frequency grid
+    assume_regular_frequency : bool (default = False)
+        if True, then do not check whether frequency is a regular grid
+
+    Returns
+    -------
+    f0, df, N : scalars
+        Parameters such that all(frequency == f0 + df * np.arange(N))
+    """
+    frequency = np.asarray(frequency)
+    if frequency.ndim != 1:
+        raise ValueError("frequency grid must be 1 dimensional")
+    elif len(frequency) == 1:
+        return frequency[0], frequency[0], 1
+    elif not assume_regular_frequency:
+        diff = frequency[1:] - frequency[:-1]
+        if not np.allclose(diff[0], diff):
+            raise ValueError("frequency must be a regular grid")
+
+    return frequency[0], frequency[1] - frequency[0], len(frequency)
+
+
+def _is_regular(frequency, assume_regular_frequency=False):
+    if assume_regular_frequency:
+        return True
+
+    frequency = np.asarray(frequency)
+
+    if frequency.ndim != 1:
+        return False
+    elif len(frequency) == 1:
+        return True
+    else:
+        diff = frequency[1:] - frequency[:-1]
+        return np.allclose(diff[0], diff)
+
+
+def validate_method(method, dy, fit_bias, nterms,
+                    frequency, assume_regular_frequency):
+    methods = available_methods()
+    fast_method_ok = ('fast' in methods)
+    prefer_fast = (fast_method_ok and len(frequency) > 200
+                   and _is_regular(frequency, assume_regular_frequency))
```

Comment:
So `and (assume_regular_frequency or _is_regular(frequency))`
(`_is_regular` will not get executed if `assume_regular_frequency` holds, since python short-circuits these expressions).